Define sanitization for db.query.text (#1100)
open-telemetry/semantic-conventions: docs/database/database-spans.md

If a corresponding `{target}` value is not available for a specific operation, the instrumentation SHOULD omit the `{target}`.
For example, for an operation describing SQL query on an anonymous table like `SELECT * FROM (SELECT * FROM table) t`, span name should be `SELECT`.

> ## Sanitization of `db.query.text`

Reviewer: Can it be moved down to be under the attribute table? Then we'll first introduce the attribute and then talk about its sanitization.

Author: Makes sense, moved this part below.

> Sanitization MUST be done by replacing all literals with a placeholder value. Such literals

Reviewer: It could be interpreted as that sanitization is a MUST and I hope we can clarify it with something like

```suggestion
The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.

Sanitization SHOULD replace all literals with a placeholder value. Such literals
```

I intentionally changed MUST to SHOULD since the sanitization so far meant removing sensitive information, not all literals. Happy to discuss more.

But also it feels that we use MUST to tell **how** to do sanitization and it's too strong.

Author: `SHOULD` sounds ok to me here as well. I was going back and forth between `SHOULD` and `MUST` - so let's just move on with `SHOULD`.

**[9]:** If using a port other than the default port for this DBMS and if `server.address` is set.

**[10]:** For sanitization see [Sanitization of `db.query.text`](../../docs/database/database-spans.md#sanitization-of-dbquerytext)
For batch operations, if the individual operations are known to have the same query text then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated with separator `; ` or some other database system specific separator if more applicable.

> **[11]:** SHOULD be collected by default only if there is sanitization that excludes sensitive information.

Reviewer: let's reference the new section here instead of the current text in the note?

When additional attributes are added that only apply to a specific DBMS, its identifier SHOULD be used as a namespace in the attribute key as for the attributes in the sections below.

## Sanitization of `db.query.text`

The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.
Sanitization SHOULD replace all literals with a placeholder value.

> Such literals include, but are not limited to, String literals, Numeric literals, Date and Time literals,
> Boolean literals, Interval literals, Binary literals, and Hexadecimal literals.

Reviewer: super-nit: 

maybe
```suggestion
Such literals include, but are not limited to, String, Numeric, Date and Time,
Boolean, Interval, Binary, and Hexadecimal literals.
```

The placeholder value SHOULD be `?`, unless it already has a defined meaning in the given database system,
in which case the instrumentation MAY choose a different placeholder.

> Placeholders in a parameterized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is,
> and there is no need to sanitize the text `$1`.

Reviewer: nit:

```suggestion
Placeholders in a parameterized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is.
```

When additional attributes are added that only apply to a specific DBMS, its identifier SHOULD be used as a namespace in the attribute key as for the attributes in the sections below.

## Sanitization of `db.query.text`

The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.
Sanitization SHOULD replace all literals with a placeholder value.
Such literals include, but are not limited to, String literals, Numeric literals, Date and Time literals,
Boolean literals, Interval literals, Binary literals, and Hexadecimal literals.
The placeholder value SHOULD be `?`, unless it already has a defined meaning in the given database system,
in which case the instrumentation MAY choose a different placeholder.

> Placeholders in a parametrized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is,

Reviewer 2: ```suggestion
Placeholders in a parameterized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is,
```

> and there is no need to sanitize the value `$1`.

Reviewer 2: ```suggestion
and there is no need to sanitize the text `$1`.
```

When additional attributes are added that only apply to a specific DBMS, its identifier SHOULD be used as a namespace in the attribute key as for the attributes in the sections below.

## Sanitization of `db.query.text`

The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.
Sanitization SHOULD replace all literals with a placeholder value.
Such literals include, but are not limited to, String, Numeric, Date and Time,

> Boolean, Interval, Binary, and Hexadecimal literals.

Reviewer 3: nit: you can probably remove the "literals" on the "Hexadecimal" too, since it already starts the sentence saying those are type of literals (I saw you removed from the other on your latest commit)

Author: I'm not a native speak, so I just asked AI and it says it's ok to drop it, but keeping it there is more precise. It says  `"literals" at the end makes it explicitly clear that all the items in the list are types of literals`. That's also how I'd understand it - the "literals" is not on the "Hexadecimal", but on the whole list.
